New buffer type for notmuch's named query strings

pazz opened the pull request:
This adds a new buffer that displays named queries similar to the taglist mode.
It so far offers only one command: select, which opens a new search buffer for the selected named query.

This is a rebased , fully splitted and refactored version of #1206, for easy review.
TODO:

- [x]  update the theme files in `alot/extra/themes`
- [x] check in auto-generated docs
- [x] add commands to manimulate named queries in this buffer
- [x] add command to save a query in search buffers
- [x] tab completion for named queries in prompts
- [ ] tests
To test this, add some named queries manually via the notmuch CLI:

```
notmuch config set query.foo "bar"
```

lucc commented:
I can not start alot on this branch. I get this exception:
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "alot/__main__.py", line 14, in <module>
    from alot.ui import UI
  File "/home/luc/vcs/alot/alot/ui.py", line 14, in <module>
    from .buffers import BufferlistBuffer
  File "/home/luc/vcs/alot/alot/buffers/__init__.py", line 11, in <module>
    from .namedqueries import NamedQueriesBuffer
  File "/home/luc/vcs/alot/alot/buffers/namedqueries.py", line 8, in <module>
    from ..widgets.namedqueries import QuerylineWidget
ModuleNotFoundError: No module named 'alot.widgets.namedqueries'
```

mjg commented:
The rebase lost the widget `alot/widgets/namedqueries.py`

pazz commented:
yep, sorry about that.
By the way, this special text widget to display pairs of strings might actually be a bit over the top: we can just as well set up ordinary `urwid.Text` widgets in the buffer. The same holds for `BufferlistBuffer`s which also have a dedicated Text widget. But perhaps we should refactor these together in another PR..

pazz commented:
There is more to do/fix here:
The commands to write/delete named queries need to be handeled with care because just as tag commands, they write to the notmuch database which can in fact be locked!

I'm also not entirely happy with the two getters for named queries as currently offered by the settings manager. Rather, as the notmuch bindings suggest, we can directly read the pairs as a dict from the bindings.

pazz commented:
I've refactored the changes to the database manager just now and also removed the problematic setters for now. This means at the moment, this PR contains a read-only version of the named queries functionality. It can only display queries, not create new ones.

To test this, add some named queries manually via the notmuch CLI:

```
notmuch config set query.foo "bar"
```

pazz commented:
OK, I think this is ready. For some reason travis is stuck, but the (unchanged) tests run fine locally. I could'nt be arsed to write new ones so far *cough*

pazz commented:
It seems that we did not have a testcase for the database manager at all so far. The last commits adds this, together with a basic test for the new functionality.

You will want to have a closer look at the `setupClass()` method and related calls as I don't have a good understanding of how tests should be set up properly!

lucc commented on tests/db/manager_test.py:
```diff
@@ -0,0 +1,49 @@
+
+
+class TestDBManager(utilities.TestCaseClassCleanup):
+    """Test the get_account_by_address helper."""
```
this looks like a copy paste error

pazz replied:
thanks. I just removed this line.

dcbaker commented on alot/buffers/namedqueries.py:
```diff
@@ -0,0 +1,77 @@
+        self.queries = self.ui.dbman.get_named_queries()
+
+        if self.isinitialized:
+            focusposition = self.querylist.get_focus()[1]
```
Let's set `self.querylist = None` in the constructor, please. Not only does it fix warnings like this, but it also helps my autocompleter in vim :)

dcbaker commented on alot/commands/globals.py:
```diff
@@ -553,6 +553,21 @@ def apply(self, ui):
+@registerCommand(MODE, 'namedqueries')
+class NamedQueriesCommand(Command):
+    """opens named queries buffer"""
+    def __init__(self, filtfun=lambda x: x, **kwargs):
```
this doesn't match the definition in the docstring, I think we should either replace the lambda with `bool`, (which will return True for any string that != "").

pazz replied:
I don't quite see how a callable would match type `bool`, but ok..

dcbaker replied:
what I mean is, `lambda x: x` is  a function that takes a thing `T` and returns `T`, but the documentation says that the callable is a function that takes type `str` and returns type `bool`; `bool()` is a function that takes `T` and returns `bool`, a superset of what the documentation says. In most cases the lambda will work, since any type not bool that is used with a logical operator will have it's `__bool_()` method called, but there are cases where it wouldn't work, like if someone where to write `if filtfun(val) is True: ...`.

dcbaker commented on alot/commands/globals.py:
```diff
@@ -979,3 +994,94 @@ def apply(self, ui):
+        :param alias: name to use for query string
+        :type alias: str
+        :param query: query string to save
+        :type query: str
```
:type query: str or None

dcbaker commented on alot/widgets/namedqueries.py:
```diff
@@ -0,0 +1,32 @@
+"""
+Widgets specific to Namedqueries mode
+"""
+from __future__ import absolute_import
```
Don't need this anymore, this is the default behavior of python 3.

dcbaker commented on tests/db/manager_test.py:
```diff
@@ -0,0 +1,48 @@
+        cls.addClassCleanup(os.unlink, f.name)
+
+        # define an empty notmuch database in a temporary directory
+        cls.dbpath = tempfile.mkdtemp()
```
should probably add `cls.addClassCleanup(shutil.rmtree, cls.dbpath)` here.

dcbaker replied:
Although, actually, we probably need to add `cls.addClassCleanup(cls.manager.kill_search_processes)` first, right?

pazz replied:
it can't hurt, yes

pazz commented:
Oh I see, that makes sense.
So Is it safe/OK to just use `bool` as type as it is in the updated
version? (If so, I'm finished with this series)

Quoting Dylan Baker (2018-07-24 22:15:26)
> @dcbaker commented on this pull request.
> 
> ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━
> 
> In alot/commands/globals.py:
> 
> > @@ -553,6 +553,21 @@ def apply(self, ui):
>              ui.buffer_open(buffers.TagListBuffer(ui, tags, self.filtfun))
> 
> 
> +@registerCommand(MODE, 'namedqueries')
> +class NamedQueriesCommand(Command):
> +    """opens named queries buffer"""
> +    def __init__(self, filtfun=lambda x: x, **kwargs):
> 
> what I mean is, lambda x: x is a function that takes a thing T and returns T,
> but the documentation says that the callable is a function that takes type str
> and returns type bool; bool() is a function that takes T and returns bool, a
> superset of what the documentation says. In most cases the lambda will work,
> since any type not bool that is used with a logical operator will have it's
> __bool_() method called, but there are cases where it wouldn't work, like if
> someone where to write if filtfun(val) is True: ....
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.*
>

dcbaker commented:
yeah, replacing `lambda x: x` with `bool` will be safe.

```python
>>> bool('')
False
>>> bool('Foo')
True
```

dcbaker commented:
LGTM! thanks.
